kernprof.py

```diff
@@ -383,27 +394,78 @@ def positive_float(value):
                             help="If specified, modules specified to `--prof-mod` will also autoprofile modules that they import. "
                             "Only works with line_profiler -l, --line-by-line")
 
-        if parser is help_parser or module is None:
+        if parser is help_parser or module is literal_code is None:
             parser.add_argument('script',
-                                metavar='{script | -m module}',
-                                help='The python script file or module to run')
+                                metavar='{path/to/script'
+                                ' | -m path.to.module | -c "literal code"}',
+                                help='The python script file, module, or '
+                                'literal code to run')
         parser.add_argument('args', nargs='...', help='Optional script arguments')
 
     # Hand off to the dummy parser if necessary to generate the help
     # text
     options = real_parser.parse_args(args)
     if help_parser and getattr(options, 'help', False):
         # This should raise a `SystemExit`
-        help_parser.parse_args([*args, '-m', module])
+        help_parser.parse_args([*args, '-m', 'dummy'])
     try:
         del options.help
     except AttributeError:
         pass
-    # Add in the pre-partitioned arguments cut off by `-m <module>`
+    # Add in the pre-partitioned arguments cut off by `-m <module>` or
+    # `-c <script>`
     options.args += post_args
     if module is not None:
         options.script = module
 
+    tempfile_source_and_content = None
+    if literal_code is not None:
+        tempfile_source_and_content = 'command', literal_code
+    elif options.script == '-' and not module:
+        tempfile_source_and_content = 'stdin', sys.stdin.read()
+
+    if not tempfile_source_and_content:
+        return _main(options, module)
+
+    # Importing `ast` is IIRC somewhat expensive, so don't do that if we
+    # don't have to
+    import ast
```

Comment:
It's hard to see that there is a `return _main` above here and that the `import ast` won't be a common path. Let's refactor this last part starting at the `import ast` into a separate function to keep the size of individual functions smaller and make it so the return points of each function are more closely grouped.